pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -600,6 +606,13 @@ func ResolvePipelineTask(
 		if run != nil {
 			rpt.RunObject = run
 		}
+	case rpt.PipelineTask.IsMatrixed() && containsArrayResultReferences(rpt.PipelineTask.Matrix):
+		// If the matrix contains array results references, get the singular TaskRunName. All the TaskRunNames won't be
+		// retrieved until applying TaskResults to the matrixed pipelineTask since the # of combinations is non-deterministic
+		rpt.TaskRunName = GetTaskRunName(pipelineRun.Status.ChildReferences, pipelineTask.Name, pipelineRun.Name)
+		if err := rpt.resolvePipelineRunTaskWithTaskRun(ctx, rpt.TaskRunName, getTask, getTaskRun, pipelineTask); err != nil {
+			return nil, err
+		}
```

Comment:
why do we get the singular task name for taskruns only? why not do it as we do for custom tasks? that is, instead of introducing a new case, check `containsArrayResultReferences` before `GetNamesOfTaskRuns` in the case below?

Reply:
I believe the reason is due to the fact that a customTask cannot have a reference to another regular Task.

For a regular Task, we need to `resolvePipelineRunTaskWithTaskRun()` with the Task Name or else it won't have the taskReferences associate with it.
![image](https://user-images.githubusercontent.com/46881325/236339689-2a236d4a-aaca-4940-a735-d1c345528b33.png)

A CustomTask only references the APIversion
![Screenshot 2023-05-04 at 6 06 46 PM](https://user-images.githubusercontent.com/46881325/236340023-d2c517b5-70e9-4da3-bc91-abfec136f1be.png)

@jerop  @lbernick

Reply:
I chatted with Emma offline about this, but I still find this very confusing. I asked Emma to look into whether it's possible to do all TaskRun resolution in the same place, as opposed to doing it ResolvePipelineTask for most pipeline tasks, and runNextSchedulableTask for pipeline tasks fanned out from array results. I'm concerned that special-casing array result references is going to lead to bugs. @EmmaMunley please let me know what you figure out; happy to schedule time to pair on this.